### ISSUE TITLE
questions about accessing categories and topics via the api

### ISSUE THREAD

**Issue by lwasser** [bennylope/pydiscourse#33]
HI There @bennylope !! 
this tool is really fantastic. I"m playing around with it and trying to understand the discourse api. This:

https://github.com/bennylope/pydiscourse/blob/master/pydiscourse/client.py#L481

seems to be the way to grab a list of categories and topics. 

## Questions For You

1. How do i get a list of categories AND subcategories?
i can get a list of categories using

`client.categories()`

But this doesn't return the subcategories from what i can tell.

2. Id like to get a list of topics within each category. When i run `client.category_topics(10)` (10 is the ID for the category that i'm interested in) I get:

`DiscourseError: Invalid Response, expecting "application/json; charset=utf-8" got "text/html; charset=utf-8"`

i've played with providing it the slug and the category name and it doesn't seem to work

just curious if you can help me sort this out. i'm happy to update docstrings and docs if i can figure this out if that is of interest / welcome!

**Comment by abkfenris** [comment]
`client.category_topics(int)` is the same method that I use for getting the topics in categories. 

Digging into my error logs, it looks like Discourse changed their API structure in the last week.

Now when you call `client.category_topics(10)` the client makes a request to `https://{YOUR_DISCOURSE_HOST}/c/10.json`. Currently from there Discourse is responding with a 302 redirect to `https://{YOUR_DISCOURSE_HOST}/c/{CATEGORY_SLUG}/10.json`.

pydiscourse right now [does not follow redirects](https://github.com/bennylope/pydiscourse/issues/32) meaning your request stops there and throws an exception as the 302 redirect returns a HTML page rather than a JSON document.

If you call `client.category_topics('{CATEGORY_SLUG}/10')` then it will return the topics in the category.

**Comment by lwasser** [comment]
ah ha! @abkfenris thank you... making progress. I have now two related questions for you in case you happen to have run into this.

Your solution above does indeed work. I now am able to generate a list of topics within a master category - so this includes all of the subtopics. 

However, when i attempt to summarize them i notice that all topics in the most recent SUBCATEGORY are listed. However there are topics MISSING in the older / less recent subcategories (i define less recent based upon creation date and when posts occurred - these are student posts so they occur in particular time periods). 

Question: is it possible that there is some sort of pagination or number of results returned that may be preventing the call to return ALL posts in a category? i definitely see a trend towards the most recent posts being returned. and older ones are missing

question 2: a work around could be to grab stats for each subtopic individually however, when i try to access `client.category_topics({CATEGORY_SLUG}/19)` where 19 is a subcategory within category 10 , i get another error.

```
DiscourseError: Invalid Response, expecting "application/json; charset=utf-8" got "text/html; charset=utf-8"
```

NOTE: i did read somewhere that the call for subcategories + categories had changed or was different. i'm finding the api documentation to be a bit tricky to follow and am then stalking various help posts online.

many thanks for any suggestions!

**Comment by abkfenris** [comment]
__1__: It looks like there is [pagination](https://docs.discourse.org/#tag/Categories/paths/~1c~1{id}.json/get) going on. You should be able to specify `client.category_topics(10, page=2)` and so forth to access more topics.

__2__: Ah, you need the whole path of slugs, so `client.category_topics('{PARENT_CATEGORY_SLUG}/{CATEGORY_SLUG}/10')`.

**Comment by lwasser** [comment]
hi again @abkfenris !! this is definitely helpful.  Have you found a way to actually get a list of category (parent and sub category) slugs?  

So i now have a full list of topics by category which is great. And i have the category ID associated with both categories and subcategories! BUT i can't seem to find the subcategory slugs to automate a full workflow of grabbing topics within each category / subcategory. The missing piece is the sub category slug that you referenced above.  thank you AGAIN.

**Comment by abkfenris** [comment]
I haven't figured out a route to find the subcategory slugs programmatically myself. If you visit the subcategory in the browser it should give you `/{parent_slug}/{sub_slug}/{sub_id}` that you can use for querying it now.

Once #34 lands and is released, you'll be able to go back to querying by `id` only.

**Comment by bennylope** [comment]
#34 has been released as version 1.1.0
